mypy/checkmember.py

```diff
@@ -368,33 +368,23 @@ def lookup_member_var_or_accessor(info: TypeInfo, name: str,
         return None
 
 
-def check_method_type(functype: FunctionLike, itype: Instance, is_classmethod: bool,
+def check_method_type(functype: FunctionLike, original_type: Type, is_classmethod: bool,
                       context: Context, msg: MessageBuilder) -> None:
     for item in functype.items():
         if not item.arg_types or item.arg_kinds[0] not in (ARG_POS, ARG_STAR):
             # No positional first (self) argument (*args is okay).
-            msg.invalid_method_type(item, context)
-        elif not is_classmethod:
-            # Check that self argument has type 'Any' or valid instance type.
-            selfarg = item.arg_types[0]
-            # If this is a method of a tuple class, correct for the fact that
-            # we passed to typ.fallback in analyze_member_access. See #1432.
-            if isinstance(selfarg, TupleType):
-                selfarg = selfarg.fallback
-            if not subtypes.is_subtype(selfarg, itype):
-                msg.invalid_method_type(item, context)
+            msg.fail('Attribute function of type %s does not accept self argument'
+                     % msg.format(item), context)
         else:
-            # Check that cls argument has type 'Any' or valid class type.
+            # Check that self argument has type 'Any' or valid instance/class type.
+            selfarg = item.arg_types[0]
             # (This is sufficient for the current treatment of @classmethod,
             # but probably needs to be revisited when we implement Type[C]
             # or advanced variants of it like Type[<args>, C].)
-            clsarg = item.arg_types[0]
-            if isinstance(clsarg, CallableType) and clsarg.is_type_obj():
```

Comment:
It's because classmethod is handled differently, falling from the instance to the type (instead of the other way around, bullet (4) above). I think the comment is not needed too.

Actually, I think the entire function can be merged with `bind_self()` since it performs overlapping checks.